library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -0,0 +1,594 @@
+    }
+  }
+
+  /**
+   * A FileSystem Wal Reader
+   */
+  public static class FileSystemWALReader implements WAL.WALReader<FileSystemWALPointer>
+  {
+    private FileSystemWALPointer currentPointer = new FileSystemWALPointer(0, 0);
+
+    private transient DataInputStream inputStream;
+    private transient Path currentOpenPath;
+
+    private final FileSystemWAL fileSystemWAL;
+    private transient FileContext fileContext;
+
+    private FileSystemWALReader()
+    {
+      //for kryo
+      fileSystemWAL = null;
+    }
+
+    public FileSystemWALReader(@NotNull FileSystemWAL fileSystemWal)
+    {
+      this.fileSystemWAL = Preconditions.checkNotNull(fileSystemWal, "wal");
+    }
+
+    protected void open(@NotNull FileContext fileContext) throws IOException
+    {
+      this.fileContext = Preconditions.checkNotNull(fileContext, "fileContext");
+    }
+
+    protected void close() throws IOException
+    {
+      if (inputStream != null) {
+        inputStream.close();
+        inputStream = null;
+      }
+    }
+
+    @Override
+    public void seek(FileSystemWALPointer pointer) throws IOException
+    {
+      if (inputStream != null) {
+        close();
+      }
+      inputStream = getInputStream(pointer);
+      Preconditions.checkNotNull(inputStream, "invalid pointer " + pointer);
+      currentPointer = pointer;
+    }
+
+    /**
+     * Move to the next WAL segment.
+     *
+     * @return true if the next part file exists and is opened; false otherwise.
+     * @throws IOException
+     */
+    private boolean nextSegment() throws IOException
+    {
+      if (inputStream != null) {
+        close();
+      }
+
+      currentPointer = new FileSystemWALPointer(currentPointer.partNum + 1, 0);
+      inputStream = getInputStream(currentPointer);
+
+      return inputStream != null;
+    }
+
+    private DataInputStream getInputStream(FileSystemWALPointer walPointer) throws IOException
+    {
+      Preconditions.checkArgument(inputStream == null, "input stream not null");
+      Path pathToReadFrom;
+      String tmpPath = fileSystemWAL.tempPartFiles.get(walPointer.getPartNum());
+      if (tmpPath != null) {
+        pathToReadFrom = new Path(tmpPath);
+      } else {
+        pathToReadFrom = new Path(fileSystemWAL.getPartFilePath(walPointer.partNum));
+      }
+
+      LOG.debug("path to read {} and pointer {}", pathToReadFrom, walPointer);
+      if (fileContext.util().exists(pathToReadFrom)) {
+        DataInputStream stream = fileContext.open(pathToReadFrom);
+        if (walPointer.offset > 0) {
+          stream.skip(walPointer.offset);
+        }
+        currentOpenPath = pathToReadFrom;
+        return stream;
+      }
+      return null;
+    }
+
+    @Override
+    public Slice next() throws IOException
+    {
+      do {
+        if (inputStream == null) {
+          inputStream = getInputStream(currentPointer);
+        }
+
+        if (inputStream != null && !fileContext.util().exists(currentOpenPath)) {
```

Comment:
This is very expensive. Why not invalidate the reader when the file rename occurs?